### PR TITLE
[MPS] Add Metal implementation of exp op

### PR DESCRIPTION
**Pull request by malfet**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #128421
* #128375
* #128373

To improve accuracy, use `precise::exp()` (and `precise::sin()`/`precise::cos()` for complex flavor)
Reuse `test_exp1` to check that accuracy of `exp` ops is sometimes closer to CPU

Fix bug in non-contiguous tensors handling

Fixes https://github.com/pytorch/pytorch/issues/84936

### CODE REVIEW

File: aten/src/ATen/native/mps/UnaryConstants.h [hunk 1]
```diff
@@ -35,4 +35,18 @@ kernel void erfinv_mps_kernel( device {0} *output [[buffer(0)]],
   }}
 
   output[index] = {0}(x);
-}})METAL";
+}}
+
+kernel void exp_kernel( device {0} *output [[buffer(0)]],
```
Comment (kulinseth) [inline aten/src/ATen/native/mps/UnaryConstants.h:40]: Can you add tanh as well ?

**Comment by kulinseth** [comment]
@malfet , can we please add a test to make sure exp and tanh are precise.

File: aten/src/ATen/native/mps/UnaryConstants.h (hunk 1 above)
Comment (malfet) [reply aten/src/ATen/native/mps/UnaryConstants.h:40]: Sure, added, those complex implementation gets a bit hairy...

**Comment by malfet** [comment]
Can we please add a test to make sure exp and tanh are precise.

@kulinseth , do you have some examples in mind? I.e. see this comment https://github.com/pytorch/pytorch/issues/84936#issuecomment-2142220367 exp on CPU and MPS do not match, even if one uses high-precision variants

**Review by kulinseth — APPROVED** [review]
Looks good.

**Pull request closed without merging**